src/main/lrsql/ops/query/reaction.clj

```diff
@@ -0,0 +1,163 @@
+   :error       {:type    error-type
+                 :message error-message}})
+
+(defn- reaction-query
+  [bk tx ruleset reaction-id trigger-id trigger-stored statement-identity]
+  (try
+    [true (ur/query-reaction
+           bk
+           tx
+           {:ruleset            ruleset
+            :trigger-id         trigger-id
+            :trigger-stored     trigger-stored
+            :statement-identity statement-identity})]
+    (catch Exception ex
+      (log/errorf
+       ex
+       "Reaction Query Error - Reaction ID: %s"
+       reaction-id)
+      [false (reaction-error-response
+              reaction-id
+              trigger-id
+              "ReactionQueryError"
+              (ex-message ex))])))
+
+(defn- generate-statement
+  [reaction-id trigger-id ruleset-match template]
+  (try
+    [true (ru/generate-statement
+           ruleset-match
+           template)]
+    (catch Exception ex
+      (log/errorf
+       ex
+       "Reaction Template Error - Reaction ID: %s"
+       reaction-id)
+      [false (reaction-error-response
+              reaction-id
+              trigger-id
+              "ReactionTemplateError"
+              (ex-message ex))])))
+
+(defn query-statement-reactions
+  "Given a statement ID, produce any reactions to that statement."
+  [bk tx {:keys [trigger-id]}]
+  (let [active-reactions (ur/query-active-reactions bk tx)]
+    ;; If there are no reactions, short-circuit w/o additional queries
+    (if (empty? active-reactions)
+      {:result []}
+      (let [{statement :payload}  (bp/-query-statement
+                                   bk tx {:statement-id trigger-id})
+            {s-reactions :result} (ur/query-reaction-history
+                                   bk tx {:statement-id trigger-id})]
+        {:result
+         (into
+          []
+          (mapcat
+           (fn [{:keys       [ruleset]
+                 reaction-id :id}]
+             ;; Cycle Check
+             (if (contains? s-reactions reaction-id)
+               (do
+                 (log/warnf
+                  "Reaction %s found in statement %s history, ignoring!"
+                  reaction-id trigger-id)
+                 []) ;; ignore
+               (let [{:keys [identityPaths
+                             template]} ruleset
+                     statement-identity (ru/statement-identity
+                                         identityPaths statement)]
+                 (if-not statement-identity
+                   [] ;; ignore
+                   (let [stored (u/str->time (get statement "stored"))
+                         [q-success ?q-result-or-error]
+                         (reaction-query
+                          bk tx ruleset reaction-id trigger-id stored
+                          statement-identity)]
+                     (if (false? q-success)
+                       ;; Query Error
+                       [?q-result-or-error]
+                       (for [ruleset-match ?q-result-or-error
```

Comment:
Side note but this pattern (as well as the similar one with `?t-result-or-error` below) can be refactored using an [exception monad](https://medium.com/@huund/monadic-error-handling-1e2ce66e3810). That said, it's _probably_ overkill for the time being though.

Reply:
This pattern `[success? result-or-error]` is basically the poor man's exception monad, lol